File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/hierarchy/calls/CalleeReferenceVisitorBase.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package org.jetbrains.kotlin.idea.hierarchy.calls;

import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.kotlin.descriptors.DeclarationDescriptor;
import org.jetbrains.kotlin.descriptors.FunctionDescriptor;
import org.jetbrains.kotlin.descriptors.PropertyDescriptor;
import org.jetbrains.kotlin.psi.*;
import org.jetbrains.kotlin.resolve.BindingContext;
import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils;

public abstract class CalleeReferenceVisitorBase extends KtTreeVisitorVoid {
    private final BindingContext bindingContext;
    private final boolean deepTraversal;

    protected CalleeReferenceVisitorBase(BindingContext bindingContext, boolean deepTraversal) {
        this.bindingContext = bindingContext;
        this.deepTraversal = deepTraversal;
    }

    protected abstract void processDeclaration(@NotNull KtSimpleNameExpression reference, @NotNull PsiElement declaration);

    @Override
    public void visitKtElement(@NotNull KtElement element) {
        if (deepTraversal || !(element instanceof KtClassOrObject || element instanceof KtNamedFunction)) {
            super.visitKtElement(element);
        }
    }

    @Override
    public void visitSimpleNameExpression(@NotNull KtSimpleNameExpression expression) {
        DeclarationDescriptor descriptor = bindingContext.get(BindingContext.REFERENCE_TARGET, expression);
        if (descriptor == null) return;

        PsiElement declaration = DescriptorToSourceUtils.descriptorToDeclaration(descriptor);
        if (declaration == null) return;

        if (isProperty(descriptor, declaration) || isCallable(descriptor, declaration, expression)) {
            processDeclaration(expression, declaration);
        }
    }

    // Accept callees of JetCallElement which refer to Kotlin function, Kotlin class or Java method
    private static boolean isCallable(DeclarationDescriptor descriptor, PsiElement declaration, KtSimpleNameExpression reference) {
        KtCallElement callElement = PsiTreeUtil.getParentOfType(reference, KtCallElement.class);
        if (callElement == null || !PsiTreeUtil.isAncestor(callElement.getCalleeExpression(), reference, false)) return false;

        return descriptor instanceof FunctionDescriptor
                 && (declaration instanceof KtClassOrObject
                     || declaration instanceof KtNamedFunction
                     || declaration instanceof PsiMethod);
    }

    // Accept only properties (not local variables or references to Java fields)
    private static boolean isProperty(DeclarationDescriptor descriptor, PsiElement declaration) {
        return descriptor instanceof PropertyDescriptor && declaration instanceof KtProperty;
    }
}

File: plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/hierarchy/calls/CalleeReferenceVisitorBase.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.jetbrains.kotlin.idea.k2.codeinsight.hierarchy.calls

import com.intellij.psi.PsiElement
import com.intellij.psi.PsiMethod
import com.intellij.psi.util.PsiTreeUtil
import org.jetbrains.kotlin.idea.references.mainReference
import org.jetbrains.kotlin.psi.KtCallElement
import org.jetbrains.kotlin.psi.KtClassOrObject
import org.jetbrains.kotlin.psi.KtElement
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.psi.KtNamedFunction
import org.jetbrains.kotlin.psi.KtProperty
import org.jetbrains.kotlin.psi.KtSimpleNameExpression
import org.jetbrains.kotlin.psi.KtTreeVisitorVoid

abstract class CalleeReferenceVisitorBase protected constructor(private val deepTraversal: Boolean) : KtTreeVisitorVoid() {

    protected abstract fun processDeclaration(reference: KtSimpleNameExpression, declaration: PsiElement)

    override fun visitKtElement(element: KtElement) {
        if (deepTraversal || !(element is KtClassOrObject || element is KtNamedFunction)) {
            super.visitKtElement(element)
        }
    }

    override fun visitSimpleNameExpression(expression: KtSimpleNameExpression) {
        val declaration = expression.mainReference.resolve()
        if (declaration == null || (declaration.containingFile as? KtFile)?.isCompiled == true) return

        if (declaration is KtProperty && !declaration.isLocal || isCallable(declaration, expression)) {
            processDeclaration(expression, declaration)
        }
    }

    companion object {
        // Accept callees of KtCallElement which refer to Kotlin function, Kotlin class or Java method
        private fun isCallable(declaration: PsiElement, reference: KtSimpleNameExpression): Boolean {
            val callElement = PsiTreeUtil.getParentOfType(reference, KtCallElement::class.java)
            if (callElement == null || !PsiTreeUtil.isAncestor(callElement.calleeExpression, reference, false)) return false

            return declaration is KtClassOrObject
                    || declaration is KtNamedFunction
                    || declaration is PsiMethod
        }
    }
}
